dot/state/epoch.go

```diff
@@ -404,26 +380,36 @@ func (s *EpochState) getConfigDataFromDatabase(epoch uint64) (*types.ConfigData,
 	return info, nil
 }
 
-// getConfigDataFromMemory retrieves the BABE config data for a given epoch that belongs to the header parameter
-func (s *EpochState) getConfigDataFromMemory(epoch uint64, header *types.Header) (*types.ConfigData, error) {
-	s.nextConfigDataLock.RLock()
-	defer s.nextConfigDataLock.RUnlock()
+func retrieveFromMemory[T types.NextEpochData | types.NextConfigData](nextEpochMap map[uint64]map[common.Hash]T,
```

Comment:
Did you consider using a type for `nextEpochMap`? 

```
type nextEpochMap[T types.NextEpochData | types.NextConfigData] map[uint64]map[common.Hash]T

func (nem nextEpochMap[T]) Retrieve(blockState *BlockState, epoch uint64, header *types.Header) (*T, error) {
	atEpoch, has := nem[epoch]
	if !has {
		return nil, fmt.Errorf("%w: %d", ErrEpochNotInMemory, epoch)
	}

	headerHash := header.Hash()
	for hash, value := range atEpoch {
		isDescendant, err := blockState.IsDescendantOf(hash, headerHash)

		// sometimes while moving to the next epoch is possible the header
		// is not fully imported by the blocktree, in this case we will use
		// its parent header which migth be already imported.
		if errors.Is(err, blocktree.ErrEndNodeNotFound) {
			parentHeader, err := blockState.GetHeader(header.ParentHash)
			if err != nil {
				return nil, fmt.Errorf("cannot get parent header: %w", err)
			}

			return retrieveFromMemory(nem, blockState, epoch, parentHeader)
		}

		if err != nil {
			return nil, fmt.Errorf("cannot verify the ancestry: %w", err)
		}

		if isDescendant {
			return &value, nil
		}
	}

	return nil, fmt.Errorf("%w: %s", errHashNotInMemory, headerHash)
}
```

I took a stab and pushed to another branch: https://github.com/ChainSafe/gossamer/pull/2633/files

Reply:
Oh, I didn't, this is really interesting!